system-helper/flatpak-system-helper.c

```diff
@@ -46,11 +46,63 @@ static gboolean no_idle_exit = FALSE;
 typedef PolkitAuthorizationResult AutoPolkitAuthorizationResult;
 typedef PolkitDetails             AutoPolkitDetails;
 typedef PolkitSubject             AutoPolkitSubject;
+typedef struct _FlatpakSystemHelperOngoingPull FlatpakSystemHelperOngoingPull;
 
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+struct _FlatpakSystemHelperOngoingPull
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+
+  const gchar *commit_id;
+  const gchar *ref;
+  guint watch_id;
+  uid_t uid;
+  guint64 current_timestamp;
+
+  gchar *root_folder_name;
+  gchar *user_folder_name;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  GSubprocess *bindfs;
+};
+
+static void
+system_helper_ongoing_pull_free (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_autoptr(GFile) root_folder = g_file_new_for_path (pull->root_folder_name);
+  g_autoptr(GFile) user_folder = g_file_new_for_path (pull->user_folder_name);
+
+  flatpak_rm_rf (root_folder, NULL, NULL);
+  flatpak_rm_rf (user_folder, NULL, NULL);
+  g_free (pull->root_folder_name);
+  g_free (pull->user_folder_name);
+
+  if (pull->watch_id)
+    g_source_remove(pull->watch_id);
+  if (pull->bindfs)
+    {
+      umount (pull->user_folder_name);
+      g_subprocess_force_exit (pull->bindfs);
+      g_clear_object (&pull->bindfs);
+    }
+  if (pull->mount_monitor_id)
+    {
+      g_signal_handler_disconnect (pull->mount_monitor, pull->mount_monitor_id);
+      pull->mount_monitor_id = 0;
+    }
+  g_object_unref (pull->mount_monitor);
+
+  g_ptr_array_remove (ongoing_pulls, pull);
+  g_ptr_array_unref (ongoing_pulls);
+  if (ongoing_pulls->len == 0)
```

Comment:
Question:
... And live where?

Sure, the unreffing should occur inside the `if` block (my approach here is that, the last ongoing_pull should `unref` the `ongoing_pulls` and set it NULL. If there are other ongoing pulls, (ongoing_pulls->len > 0), we shouldn't.

Also, after reading the `g_ptr_array_free` documentation, I think having a `g_ptr_array_ref/unref` pair approach is what is desirable here. Thoughts?

Reply:
I suggest allocating the array at the start of main(), and unreffing it at the end, rather than trying to arrange for it to be `NULL` when it would otherwise be empty.

Thinking about this a bit more, and reading the code: the system helper currently exits 10 minutes after it is launched. If an unprivileged pull takes longer than 10 minutes, the helper will exit with the bind still set up. When the app comes to call Deploy(), the helper will be relaunched but will have no idea that it created the bindfs for this user. I think we need to keep the daemon alive for as long as there are any active bindfs mounts.

Reply:
I have tried to address this bit in https://github.com/endlessm/flatpak/pull/115/commits/15b016e99e838a6d43d64b50212a6e36e482243f

Thanks for the suggestion.